pkg/eth/tx_listener.go

```diff
@@ -88,11 +95,23 @@ func (t *TxListener) PollForTxCompletion(txID string, updates chan<- string) {
 
 }
 
-func (t *TxListener) checkTx(hash common.Hash) (bool, error) {
+func (t *TxListener) waitForTimeout() {
+	c := time.After(t.timeout)
+	<-c
+	t.mutex.Lock()
+	t.timeoutPassed = true
+	t.mutex.Unlock()
+}
 
+func (t *TxListener) checkTx(hash common.Hash) (bool, error) {
+	t.mutex.Lock()
+	defer t.mutex.Unlock()
 	_, isPending, err := t.blockchain.TransactionByHash(context.Background(), hash)
 	if err != nil {
 		log.Errorf("Error retrieving TransactionByHash: err: %v\n", err)
+		if err == ethereum.NotFound && !t.timeoutPassed {
```

Comment:
imo this function shouldn't be concerned whether the timeout has passed - we can just check if ethereum.NotFound in `PollForTxCompletion`

Reply:
i see what dan means here, we can move this line up into PollForTxCompletion.

Reply:
I kind of disagree, in that case your setting isPending and then flipping it back  in the parent function.  Makes the logic in the parent way more complex

Reply:
`timeoutPassed` should not be part of the struct - since that means you need to create a `TxListener` for every transaction you want to listen to. And in turn, inject all of the dependencies. Also, the way this stands it is lying - the transaction is not found, not pending. What if we wanted to check if a transaction exist?

Reply:
you could pass the information in here, and not need the struct level information.  We can't tell if a transaction exists or not immediately so i think the state of pending is true in that case.  Unless you know exactly when you expect it was created, you'd still want to wait.